core/domain/user_domain.py

```diff
@@ -107,3 +111,133 @@ def to_dict(self):
             'mute_feedback_notifications': self.mute_feedback_notifications,
             'mute_suggestion_notifications': self.mute_suggestion_notifications
         }
+
+
+class ExpUserLastPlaythrough(object):
+    """Domain object for an exploration last playthrough model."""
+
+    def __init__(self, user_id, exploration_id, version_last_played,
+                 time_last_played_msec, last_state_played):
+        self.id = '%s.%s' % (user_id, exploration_id)
+        self.user_id = user_id
+        self.exploration_id = exploration_id
+        self.version_last_played = version_last_played
+        self.time_last_played_msec = time_last_played_msec
+        self.last_state_played = last_state_played
+
+    def update_last_played_information(self, time_last_played_msec,
+                                       version_last_played,
+                                       last_state_played):
+        """Updates the last playthrough information of the user.
+
+        Args:
+            time_last_played_msec: float. The time in milliseconds since the
+                Epoch when the user last played the explorations.
+            version_last_played: int. The version of the exploration that was
+                played by the user.
+            last_state_played: str. The name of the state at which the learner
+                left the exploration.
+        """
+        self.time_last_played_msec = time_last_played_msec
+        self.version_last_played = version_last_played
+        self.last_state_played = last_state_played
+
+class IncompleteExplorations(object):
+    """Domain object for the incomplete explorations model."""
+
+    def __init__(self, user_id, incomplete_exploration_ids):
+        self.id = user_id
+        self.incomplete_exploration_ids = incomplete_exploration_ids
+
+    def add_exploration_id(self, exploration_id):
+        """Adds the exploration id to the list of incomplete exploration ids."""
+
+        self.incomplete_exploration_ids.append(exploration_id)
+
+    def remove_exploration_id(self, exploration_id):
+        """Removes the exploration id from the list of incomplete exploration
+        ids.
+        """
+
+        self.incomplete_exploration_ids.remove(exploration_id)
+
+    def get_last_playthrough_information(self, exploration_id):
+        """Get the information regarding the last playthrough of the user of
+        the given exploration.
+
+        Args:
+            exploration_id: str. The id of the exploration for which we have to
+            get the last playthrough information.
+
+        Returns:
+            ExpUserLastPlaythrough. A ExpUserLastPlaythrough domain object
+            corresponding to the given exploration id.
+        """
+        incomplete_exploration_user_model = (
```

Comment:
Here and below: domain objects should be plain Python objects and should not refer to datastore models. Such functions should be in the *_services.py file.